NDTensors/src/default_storage.jl

```diff
@@ -0,0 +1,14 @@
+## This is a fil which specifies the default storage type provided some set of parameters
+## The parameters are the element type and storage type
+default_datatype(eltype::Type{<:Number}) = Vector{eltype}
+default_eltype() = Float64
+
+default_storagetype(datatype::Type{<:AbstractArray{ElT}}) where {ElT} = Dense{ElT,datatype}
```

Comment:
`default_storagetype(datatype::Type{<:AbstractArray}) = Dense{eltype(datatype),datatype}`

Reply:
For example when we incorporate indices we might be:
```julia
functino default_storagetype(datatype::Type{<:AbstractArray{ElT}}, inds::Tuple) where {ElT}
  return set_datatype(default_storagetype(inds), datatype)
end
```
where `default_storagetype(::Tuple)` then analyzes the indices for block indices or dense indices and returns `BlockSparse` or `Dense`.

Reply:
Then this case could be a special case where:
```julia
function default_storagetype(datatype::Type{<:AbstractArray{ElT}}) where {ElT}
  return default_storagetype(datatype, ())
end
```
and empty indices defaults to `Dense`.

Reply:
Okay I think I may have figured out the answer to this `{ElT}` problem. We can use the `set_eltype_if_unspecified` function to set the eltype of the specific vectors. Then we only need one 
``` 
Function default_storagetype(datatype::Type{<:AbstractArray} inds::Tuple)
  datatype = set_eltype_if_unspecified(datatype, default_eltype())
  if typeof(inds) == Integer || length(inds) == 0
    return Dense{eltype(datatype), datatype)
  else
    return #blocksparse 
 end
 ```

Reply:
Yeah, that's seem like a good pattern to make a function `set_eltype_if_unspecified` (I guess that isn't defined right now, correct)? That should be useful in other places as well. We could also have a shorthand:
```julia
set_default_eltype_if_unspecified(datatype) = set_eltype_if_unspecified(datatype, default_eltype())
```

I could imagine this could be based on a special adaptor as part of the future `adapt_type` function, i.e.:
```julia
set_eltype_if_unspecified(datatype::Type, eltype::Type) = adapt_type(SetIfUnspecifiedAdaptor(eltype), datatype)
```

However to clarify we should find a better design based on something like `default_storagetype(inds::Tuple)` which dispatches on the `inds` rather than `if` statements to choose the storage type.

Reply:
`set_eltype_if_unspecified` is defined in `NDTensors/src/adapt.jl`.  I will write the `set_eltype_if_unspecified(datatype) = set_eltype_if_unspecified(datatype, default_eltype())` function but I don't know yet if it will be useful. We will see!

I agree with the dispatches. I don't really know what Dense vs Blocksparse inds look like so I don't know how to implement those different dispatches.

Reply:
Ah I forgot that I had already started to define `set_eltype_if_unspecified`. Looks like a good start but we'll need to generalize that functionality a bit.

No worries, we can handle that in a future PR that extends this stuff to `BlockSparse`. The designs we're developing here should help a lot once we get to that!

Reply:
I have already pushed some updates to `set_eltype_if_unspecified` to work for a `AbstractArray` and to use `similartype`